nginx/nginx.go

```diff
@@ -349,60 +350,99 @@ func (lb *nginxLoadBalancer) createConfig(update controller.IngressUpdate) ([]by
 	return output.Bytes(), nil
 }
 
-func (lb *nginxLoadBalancer) getNginxLogHeaders() string {
+func (n *nginxUpdater) getNginxLogHeaders() string {
 	headersString := ""
-	for _, nginxLogHeader := range lb.LogHeaders {
+	for _, nginxLogHeader := range n.LogHeaders {
 		headersString = headersString + " " + nginxLogHeader + "=$http_" + strings.Replace(nginxLogHeader, "-", "_", -1)
 	}
 
 	return headersString
 }
 
-type pathSet map[string]struct{}
+type upstreams []*upstream
 
-func createNginxEntries(update controller.IngressUpdate) []*nginxEntry {
-	sortedIngressEntries := update.SortedByName().Entries
-	hostToNginxEntry := make(map[string]*nginxEntry)
-	hostToPaths := make(map[string]pathSet)
-	var nginxEntries []*nginxEntry
-	var upstreamIndex int
+func (u upstreams) Len() int           { return len(u) }
+func (u upstreams) Less(i, j int) bool { return u[i].ID < u[j].ID }
+func (u upstreams) Swap(i, j int)      { u[i], u[j] = u[j], u[i] }
 
-	for _, ingressEntry := range sortedIngressEntries {
-		nginxPath := createNginxPath(ingressEntry.Path)
-		upstream := upstream{
-			ID:     fmt.Sprintf("upstream%03d", upstreamIndex),
+func createUpstreamEntries(update controller.IngressUpdate) []*upstream {
+	idToUpstream := make(map[string]*upstream)
+
+	for _, ingressEntry := range update.Entries {
+		upstream := &upstream{
+			ID:     upstreamID(ingressEntry),
 			Server: fmt.Sprintf("%s:%d", ingressEntry.ServiceAddress, ingressEntry.ServicePort),
 		}
+		idToUpstream[upstream.ID] = upstream
+	}
+
+	var sortedUpstreams []*upstream
+	for _, upstream := range idToUpstream {
+		sortedUpstreams = append(sortedUpstreams, upstream)
+	}
+
+	sort.Sort(upstreams(sortedUpstreams))
+	return sortedUpstreams
+}
+
+func upstreamID(e controller.IngressEntry) string {
+	return fmt.Sprintf("%s.%s.%d", e.Namespace, e.ServiceAddress, e.ServicePort)
+}
+
+type servers []*server
+
+func (s servers) Len() int           { return len(s) }
+func (s servers) Less(i, j int) bool { return s[i].ServerName < s[j].ServerName }
+func (s servers) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+
+type locations []location
```

Comment:
Yeah it was done that way because in `#createServerEntries` we need a pointer to the server struct, as we mutate its state as we iterate through the ingress entries. I agree it's a bit confusing - we could change locations to be a pointer as well.

If we make servers not a slice-to-pointer, then we'll have to write code to convert the slice by copying over all its contents, which I think is a bit crap!